stdlib/ballerina-grpc/src/main/java/org/ballerinalang/net/grpc/listener/ServerCallHandler.java

```diff
@@ -0,0 +1,313 @@
+                throw Status.Code.CANCELLED.toStatus().withDescription("call already cancelled").asRuntimeException();
+            }
+            if (!sentHeaders) {
+                call.sendHeaders(((Message) response).getHeaders());
+                sentHeaders = true;
+            }
+            call.sendMessage(response);
+        }
+
+        @Override
+        public void onError(RespT error) {
+
+            if (!sentHeaders) {
+                call.sendHeaders(((Message) error).getHeaders());
+                sentHeaders = true;
+            }
+            call.close(Status.fromThrowable(((Message) error).getError()), new DefaultHttpHeaders());
+        }
+
+        @Override
+        public void onCompleted() {
+
+            if (cancelled) {
+                throw Status.Code.CANCELLED.toStatus().withDescription("call already cancelled").asRuntimeException();
+            } else {
+                call.close(Status.Code.OK.toStatus(), new DefaultHttpHeaders());
+            }
+        }
+
+        @Override
+        public boolean isReady() {
+
+            return call.isReady();
+        }
+
+        public boolean isCancelled() {
+
+            return call.isCancelled();
+        }
+    }
+
+    /**
+     * Returns endpoint instance which is used to respond to the caller.
+     *
+     * @param responseObserver client responder instance.
+     * @return instance of endpoint type.
+     */
+    private BValue getConnectionParameter(Resource resource, StreamObserver<ResponseT> responseObserver) {
+        ProgramFile programFile = getProgramFile(resource);
+        // generate client responder struct on request message with response observer and response msg type.
+        BStruct clientEndpoint = BLangConnectorSPIUtil.createBStruct(programFile,
+                GrpcConstants.PROTOCOL_STRUCT_PACKAGE_GRPC, GrpcConstants.CALLER_ACTION);
+        clientEndpoint.addNativeData(GrpcConstants.RESPONSE_OBSERVER, responseObserver);
+        clientEndpoint.addNativeData(GrpcConstants.RESPONSE_MESSAGE_DEFINITION, methodDescriptor.getOutputType());
+
+        // create endpoint type instance on request.
+        BStruct endpoint = BLangConnectorSPIUtil.createBStruct(programFile,
+                GrpcConstants.PROTOCOL_STRUCT_PACKAGE_GRPC, GrpcConstants.SERVICE_ENDPOINT_TYPE);
+        endpoint.setRefField(0, clientEndpoint);
+        endpoint.setIntField(0, responseObserver.hashCode());
+        return endpoint;
+    }
+
+    /**
+     * Returns BValue object corresponding to the protobuf request message.
+     *
+     * @param requestMessage protobuf request message.
+     * @return b7a message.
+     */
+    private BValue getRequestParameter(Resource resource, RequestT requestMessage, boolean isHeaderRequired) {
+        if (resource == null || resource.getParamDetails() == null || resource.getParamDetails().size() > 3) {
+            throw new RuntimeException("Invalid resource input arguments. arguments must not be greater than three");
```

Comment:
Isn't this error misleading on the case of resource == null  or resource.getParamDetails() == null  as it is not greater than 3?